File: LockFreeSetImpl.java
package lab2;

import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreeSetImpl<T extends Comparable<T>> implements LockFreeSet<T> {

	private final Node<T> head;

	public LockFreeSetImpl() {
		head = new Node<>(null);
	}

	private class Node<T> {
		T value;
		AtomicMarkableReference<Node<T>> next;

		Node(T value) {
			this.value = value;
			next = new AtomicMarkableReference<Node<T>>(null, false);
		}
	}

	private class Pair<T> {
		Node<T> prev;
		Node<T> cur;

		Pair(Node<T> prev, Node<T> cur) {
			this.prev = prev;
			this.cur = cur;
		}
	}

	public boolean add(T value) {
		Node<T> toAdd = new Node<>(value);
		while (true) {
			Pair<T> pair = findPosition(value);
			if (value.equals(pair.prev.value) && !pair.prev.next.isMarked()) {
				return false;
			}
			toAdd.next = new AtomicMarkableReference<>(pair.cur, false);
			if (pair.prev.next.compareAndSet(pair.cur, toAdd, false, false)) {
				return true;
			}
		}
	}

	@Override
	public boolean remove(T value) {
		while (true) {
			Pair<T> pair = findPosition(value);
			if (!value.equals(pair.prev.value)) {
				return false;
			}
			
			Node<T> next = pair.prev.next.getReference();
			if(pair.prev.next.compareAndSet(next, next, false, true));
			return true;
		}
	}

	public boolean contains(T value) {
		Pair<T> pair = findPosition(value);
		if (pair.prev == head)
			return false;
		if (pair.prev.value.equals(value))
			return true;
		return false;
	}

	public boolean isEmpty() {
		while (true) {
			while (true) {
				Node<T> nextElem = head.next.getReference();
				if (nextElem == null)
					return true;
				if (!nextElem.next.isMarked())
					return false;
				if (!head.next.compareAndSet(nextElem, nextElem.next.getReference(), false, false)) {
					break;
				}
			}
		}
	}

	public Pair<T> findPosition(T value) {
		while (true) {
			Node<T> prev = head;
			Node<T> cur = head.next.getReference();

			while (true) {
				if (cur == null)
					return new Pair<T>(prev, null);
				if (cur.next.isMarked()) {
					if (!prev.next.compareAndSet(cur, cur.next.getReference(), false, false)) {
						break;
					}
					cur = prev.next.getReference();
					continue;
				}
				if (cur.value.compareTo(value) > 0) {
					return new Pair<>(prev, cur);
				}
				prev = cur;
				cur = cur.next.getReference();
			}
		}
	}
	
}

File: aptu/302_18/vinnichenko/2/src/main/kotlin/ru/spbau/maxim/hw2/LockFreeSetImpl.kt
package ru.spbau.maxim.hw2

import java.util.concurrent.atomic.AtomicMarkableReference

class LockFreeSetImpl<T : Comparable<T>> : LockFreeSet<T> {
    private val tail = Node<T>(null, null)
    private val head = Node<T>(null, tail)

    override fun add(value: T): Boolean {
        while (true) {
            val (prev, node) = findRange(value)

            if (compareWithNode(value, node) == 0) {
                return false
            }
            val newNode = Node(value, node)
            if (prev.next.compareAndSet(node, newNode, false, false)) {
                return true
            }
        }
    }

    override fun remove(value: T): Boolean {
        val isDeletedBuf = booleanArrayOf(false)
        while (true) {
            val (_, node: Node<T>) = findRange(value)
            val nextNode = node.next.get(isDeletedBuf)
            if (compareWithNode(value, node) != 0 || isDeletedBuf[0]) {
                return false
            }
            if (node.next.attemptMark(nextNode, true)) {
                return true
            }
        }
    }

    override fun contains(value: T): Boolean {
        val (_, node: Node<T>) = findRange(value)
        return value == node.value && !node.next.isMarked
    }

    override fun isEmpty(): Boolean {
        val (l, r) = findRangePredicate { node -> !node.next.isMarked }
        return l === head && r === tail
    }

    private fun compareWithNode(value: T, node: Node<T>): Int = when (node) {
        head -> 1
        tail -> -1
        else -> value.compareTo(node.value!!)
    }

    private class Node<T>(val value: T?, nextNode: Node<T>?) {
        val next = AtomicMarkableReference(nextNode, false)
    }

    private fun findRange(value: T): Pair<Node<T>, Node<T>> {
        return findRangePredicate { node -> compareWithNode(value, node) <= 0 }
    }

    //находит первый элемент, подходящий под predicate, и предыдуший (считается, что tail всегда подходит по predicate)
    //возвращает Pair(предыдущий, подходящий элемент)
    //также выполняет helping
    private fun findRangePredicate(predicate: (Node<T>) -> Boolean): Pair<Node<T>, Node<T>> {
        val isDeletedBuf = booleanArrayOf(false)
        loop@ while (true) {
            var prev: Node<T>? = null
            var node = head
            var nextNode = head.next.get(isDeletedBuf)
            while (nextNode != null && (node === head || !predicate(node))) {
                prev = node
                node = nextNode
                nextNode = node.next.get(isDeletedBuf)

                if (isDeletedBuf[0]) {
                    if (!prev.next.compareAndSet(node, nextNode, false, false)) {
                        continue@loop
                    } else {
                        node = prev
                    }
                }
            }

            // l < value <= r
            return Pair(prev!!, node)
        }
    }
}